### PR TITLE
Add start events

### PR DESCRIPTION
**Pull request by fcarstens**
This PR does the following:
- add `runStart`, `suiteStart`, `testStart` events
- refactor `JasmineAdapter` to emit start events during runtime
- refactor `QUnitAdapter` to emit start events during runtime
- adjust tests

Open issues: 
- Jasmine catches exceptions in `afterAll` hooks  and exposes them as `failedExpectations` on a suite. (How) do we want to handle this?

### CODE REVIEW

File: lib/Data.js [hunk 1]
```diff
@@ -30,22 +30,32 @@ export class Suite {
     }
 
     get runtime() {
-        return this.getAllTests()
+        var runtime = this.getAllTests()
             .map((x) => x.runtime)
             .reduce((sum, x) => sum + x, 0);
+        if (isNaN(runtime)) {
```
Comment (jzaefferer) [inline lib/Data.js]: Under what condition do we end up with NaN? Can we detect that condition earlier?

File: lib/adapters/JasmineAdapter.js [hunk 2]
```diff
@@ -4,68 +4,115 @@ import {Test, Suite} from "../Data.js";
...
             suiteDone: this.onSuiteDone.bind(this),
             jasmineDone: this.onJasmineDone.bind(this)
         });
-        // a stack of test arrays. The top element on the stack is the currently active suite
-        this.tests = [[]];
-        // a stack of suite arrays
-        this.suites = [[]];
+
+        // Map internal jasmine ids to standardized test/suite objects.
+        this.idObjectMap = {};
```
Comment (jzaefferer) [inline lib/adapters/JasmineAdapter.js]: This needs a better name. "object" doesn't tell the reader anything useful.

File: lib/adapters/JasmineAdapter.js [hunk 3]
```diff
@@ -4,68 +4,115 @@ import {Test, Suite} from "../Data.js";
...
+        var name = jasmineSuite.description;
+        var childSuites = [];
+        var tests = [];
+        for (let child of jasmineSuite.children) {
+            // a child can be either a suite or a spec. We use the id to distinguish between the two.
+            if (child.id.indexOf("suite") === 0) {
+                childSuites.push(this.convertSuite(child));
+            } else if (child.id.indexOf("spec") === 0) {
+                tests.push(this.convertSpec(child));
+            }
```
Comment (jzaefferer) [inline lib/adapters/JasmineAdapter.js]: There is no else here. Can the children property contain any properties that are neither suite nor spec?

File: lib/adapters/JasmineAdapter.js [hunk 4]
```diff
@@ -4,68 +4,115 @@ import {Test, Suite} from "../Data.js";
...
+
+        this.idObjectMap[jasmineSpec.id] = test;
+
+        return test;
+    }
+
+
+    onJasmineStarted() {
+        var startSuite = this.convertSuite(this.jasmine.topSuite());
+        startSuite.name = "";
```
Comment (jzaefferer) [inline lib/adapters/JasmineAdapter.js]: Why an empty string? Maybe it should be undefined instead?

File: lib/adapters/JasmineAdapter.js [hunk 5]
```diff
@@ -4,68 +4,115 @@ import {Test, Suite} from "../Data.js";
...
-            this.tests[this.tests.length - 1].push(new Test("afterAll", "failed", 0));
-        }
-        var suite = new Suite(details.description, this.suites.pop(), this.tests.pop());
-        this.emit("suiteEnd", suite);
-        this.suites[this.suites.length - 1].push(suite);
+
+        this.emit("suiteEnd", this.idObjectMap[details.id]);
+
+        /*
+         FIXME: Jasmine Suite may have failedExpectations caused by a failure in afterAll.
```
Comment (jzaefferer) [inline lib/adapters/JasmineAdapter.js]: This seems fair to ignore in the adapter, but document it as a limitation and point it out to the Jasmine team.

File: lib/adapters/QUnitAdapter.js [hunk 6]
```diff
@@ -5,18 +5,52 @@ import {Test, Suite} from "../Data.js";
 export default class QUnitAdapter extends EventEmitter {
     constructor() {
         super();
-        QUnit.done(this.onDone.bind(this));
+        QUnit.config.reorder = false;
```
Comment (jzaefferer) [inline lib/adapters/QUnitAdapter.js]: This is problematic. Is it not enough to set this for our tests here?

File: lib/adapters/QUnitAdapter.js [hunk 7]
```diff
@@ -5,18 +5,52 @@ import {Test, Suite} from "../Data.js";
...
-        QUnit.log(this.onLog.bind(this));
-        QUnit.testStart(this.onTestStart.bind(this));
+        QUnit.done(this.onDone.bind(this));
+
+        this.idObjectMap = {};
+    }
+
+    convertModule(qunitModule) {
+        var name = qunitModule.name;
+        var childSuites = [];
```
Comment (jzaefferer) [inline lib/adapters/QUnitAdapter.js]: This is used only once, can be inlined.

File: lib/adapters/QUnitAdapter.js [hunk 8]
```diff
@@ -5,18 +5,52 @@ import {Test, Suite} from "../Data.js";
...
         QUnit.moduleDone(this.onModuleDone.bind(this));
-        QUnit.log(this.onLog.bind(this));
-        QUnit.testStart(this.onTestStart.bind(this));
+        QUnit.done(this.onDone.bind(this));
+
+        this.idObjectMap = {};
+    }
+
+    convertModule(qunitModule) {
+        var name = qunitModule.name;
```
Comment (jzaefferer) [inline lib/adapters/QUnitAdapter.js]: This is also used just once. Can be replaced.

File: test/referenceData.js [hunk 9]
```diff
@@ -19,35 +18,53 @@ var globalSuiteNoNesting = new Suite("", [groupA, groupB], []);
 // Global Suite for Jasmine
 var globalSuiteNesting = new Suite("", [groupA, groupB, parent], []);
 
+function toStartTest(test) {
+    return new Test(
+        test.testName,
+        test.suiteName,
+        undefined,
+        undefined,
+        undefined
```
Comment (jzaefferer) [inline test/referenceData.js]: Explicitly passing undefined when there are no other defined arguments following seems pointless. Should just drop these three. Applies to similar usages above.

File: test/test.js [hunk 10]
```diff
@@ -10,8 +13,31 @@ jasmine.loadConfig({
     ]
 });
 
-var runner =  new JsReporters.JasmineAdapter(jasmine);
-var testReporter = new JsReporters.TestReporter(runner, referenceData.Jasmine);
+var jasmineRunner = new JsReporters.JasmineAdapter(jasmine.env);
```
Comment (jzaefferer) [inline test/test.js]: This used to pass `jasmine` instead of `jasmine.env`. Why the change?

File: test/test.js [hunk 11]
```diff
@@ -10,8 +13,31 @@ jasmine.loadConfig({
...
+var qunitTestReporter = new JsReporters.TestReporter(qunitRunner, referenceData.QUnit);
+
+QUnit.config.autorun = false;
+
+require("./qunit/tests.js");
+
+QUnit.load();
+
+done.then(function () {
+    process.exit(qunitTestReporter.ok ? 0 : 1);
```
Comment (jzaefferer) [inline test/test.js]: Might as well throw an exception here as well.

File: test/test.js [hunk 12]
```diff
@@ -10,8 +13,31 @@ jasmine.loadConfig({
...
+jasmine.addReporter({}); // Suppress the default reporter
+var jasmineTestReporter = new JsReporters.TestReporter(jasmineRunner, referenceData.Jasmine);
 
 jasmine.execute();
-process.exit(testReporter.ok ? 0 : 1);
+if (!jasmineTestReporter.ok) {
+    throw new Error("Tests for JasmineAdapter failed!");
+}
+
+//TODO: Use a proper test framework?
```
Comment (jzaefferer) [inline test/test.js]: What does that mean?

File: test/test.js [hunk 13]
```diff
@@ -10,8 +13,31 @@ jasmine.loadConfig({
...
+var done = new Promise(function (resolve) {
+    QUnit.done(function () {
+        resolve();
+    });
+});
+
+var qunitRunner = new JsReporters.QUnitAdapter();
```
Comment (jzaefferer) [inline test/test.js]: Can we pass the QUnit reference instead of relying on the global inside the adapter?

**Comment by jzaefferer** [comment]
Lots of small details to improve, but overall looking good.

File: lib/Data.js (hunk 1 above)
Comment (fcarstens) [inline lib/Data.js]: If one of the tests  has not been run yet, its runtime is undefined. If we add undefined and a number we end up with NaN. We can check all tests before, but that's probably not easier?

File: lib/adapters/JasmineAdapter.js (hunk 4 above)
Comment (fcarstens) [inline lib/adapters/JasmineAdapter.js]: This is similar to Mocha which has an empty string as the global suite title. Change anyway?

File: lib/adapters/QUnitAdapter.js (hunk 6 above)
Comment (fcarstens) [inline lib/adapters/QUnitAdapter.js]: If we have reordering enabled, QUnit may emit multiple module done events for one suite. I think we don't want to have this in the standard?

File: test/test.js (hunk 10 above)
Comment (fcarstens) [inline test/test.js]: In node, it's `jasmine.env`, in the browser it's `jasmine.getEnv()`...

File: test/test.js (hunk 11 above)
Comment (fcarstens) [inline test/test.js]: This is async, so an exception doesn't change the exit code. I added a comment to explain this.

File: lib/Data.js (hunk 1 above)
Comment (jzaefferer) [inline lib/Data.js]: Let's add a comment than, based on your description here.

File: lib/adapters/JasmineAdapter.js (hunk 4 above)
Comment (jzaefferer) [inline lib/adapters/JasmineAdapter.js]: I guess an actual name would be even better, but probably also much harder to come up with. Unless you have an idea or find a better reference, I think undefined is a better choice than the empty string.

File: lib/adapters/QUnitAdapter.js (hunk 6 above)
Comment (jzaefferer) [inline lib/adapters/QUnitAdapter.js]: Let's remove this here and keep it in the tests. And create or update a QUnit issue with details of this. It's really a QUnit bug.

File: test/test.js (hunk 11 above)
Comment (jzaefferer) [inline test/test.js]: An unhandled exception doesn't cause a non zero exit code? That's surprising, sounds like a bug in nodejs.

File: test/test.js (hunk 12 above)
Comment (fcarstens) [inline test/test.js]: Using a test framework with proper async support would make this test easier to understand. Choosing between QUnit, Jasmine, Mocha, Buster or Intern is very meta here. Worth the effort?

File: lib/adapters/QUnitAdapter.js (hunk 6 above)
Comment (fcarstens) [inline lib/adapters/QUnitAdapter.js]: Ok. Tracked at https://github.com/jquery/qunit/issues/830.

File: test/test.js (hunk 12 above)
Comment (jzaefferer) [inline test/test.js]: Now I got it. Still don't understand how an unhandled exception doesn't lead to a non zero exit code. How does the assert module work then?Afaik that can be used by itself without any extra infrastructure. 

@leobalter any idea?

File: test/test.js [hunk 14]
```diff
@@ -10,8 +13,33 @@ jasmine.loadConfig({
...
+    QUnit.done(function () {
+        resolve();
+    });
+});
+
+var qunitRunner = new JsReporters.QUnitAdapter(QUnit);
+var qunitTestReporter = new JsReporters.TestReporter(qunitRunner, referenceData.QUnit);
+
+QUnit.config.autorun = false;
+QUnit.config.reorder = false;
```
Comment (leobalter) [inline test/test.js]: This is not necessary on node envs.

File: test/test.js [hunk 15]
```diff
@@ -10,8 +13,33 @@ jasmine.loadConfig({
...
+if (!jasmineTestReporter.ok) {
+    throw new Error("Tests for JasmineAdapter failed!");
+}
+
+//TODO: Use a proper test framework?
+var done = new Promise(function (resolve) {
+    QUnit.done(function () {
+        resolve();
+    });
+});
```
Comment (leobalter) [inline test/test.js]: This might be solved as:

```js
QUnit.done(function () {
    process.exit(qunitTestReporter.ok ? 0 : 1);
});
```

File: test/test.js (hunk 12 above)
Comment (leobalter) [inline test/test.js]: I'm not sure what are we talking about, but I have some opinions from what I can see and presume:

I can't see how throwing an error on this module is better than a direct call on the process.exit if the idea is to exit. The error above on line 22 (`"Tests for JasmineAdapter failed!"`) seems weird to me.

Another issue I see on this file is that we are loading 2 reporters, one for Jasmine and one for QUnit, I wonder if they work better on individual files.

Following the file split, I'm also thinking about a proper reaction if `jasmineTestReporter.ok` or `qunitTestReporter.ok` are `false`. They can be improved to something that doesn't interrupt the tests, following to whatever comes next.

File: test/test.js (hunk 14 above)
Comment (leobalter) [inline test/test.js]: Is this related to https://github.com/jquery/qunit/issues/830 ?

Comment (fcarstens) [inline test/test.js]: Yes, but as you said, not required on node envs. I'll remove it.

File: test/test.js (hunk 12 above)
Comment (fcarstens) [inline test/test.js]: We consistently use `process.exit()` now. If we want more advanced features here, I'd suggest we use Jasmine or Mocha here (or QUnit, but Jasmine and Mocha seem to fit better for node) to structure the tests. 
Do we want to tackle this with this PR or at a later point?

**Comment by fcarstens** [comment]
I addressed all comments above. Last open discussion is https://github.com/js-reporters/js-reporters/pull/24#discussion-diff-36724127.

**Comment by jzaefferer** [comment]
Trying to look at that last issue again (still on mobile), the link doesn't work (outdated diff?). Can you post it again, here or on the updated diff?

File: lib/Data.js [hunk 16]
```diff
@@ -30,22 +30,34 @@ export class Suite {
     }
 
     get runtime() {
-        return this.getAllTests()
+        var runtime = this.getAllTests()
             .map((x) => x.runtime)
             .reduce((sum, x) => sum + x, 0);
+        // If one of the tests has not been run yet, its runtime is undefined.
```
Comment (jzaefferer) [inline lib/Data.js]: Add blank line above line comment. See if es-lint has an option for that.

File: lib/Data.js [hunk 17]
```diff
@@ -30,22 +30,34 @@ export class Suite {
...
+        // If we add undefined and a number we end up with NaN.
+        if (isNaN(runtime)) {
```
Comment (jzaefferer) [inline lib/Data.js]: How about a tertiary?

```js
return isNaN(runtime) ? undefined : runtime;
```

File: lib/adapters/QUnitAdapter.js [hunk 18]
```diff
@@ -1,22 +1,52 @@
...
+        var suite = new Suite(suiteName, [], tests);
+        this.idDataMap[qunitModule.name] = suite;
+        return suite;
+    }
 
-        this.tests = {};
-        this.suites = [];
+    onBegin() {
+        var suites = [];
+        for (let suite of this.QUnit.config.modules) {
```
Comment (jzaefferer) [inline lib/adapters/QUnitAdapter.js]: Let's add a comment here, something like "// Access QUnit internals to build liste of modules, working around missing event data"

File: test/test.js (hunk 12 above)
Comment (jzaefferer) [inline test/test.js]: I think for now its fine as is, don't need to introduce more complexity.

**Comment by jzaefferer** [comment]
Did another review of the latest changes. A few more details, then this should be good land.

**Comment by fcarstens** [comment]
Done! :smiley:

**Comment by jzaefferer** [comment]
:+1:
